File: solutions/free-pascal/all-your-base/1/AllYourBase.pas
unit AllYourBase;

{$mode ObjFPC}{$H+}

interface

type
  TIntArray = Array Of Integer;

function rebase(
  const InputBase   : integer;
  const InputDigits : TIntArray;
  const OutputBase  : integer
) : TIntArray;

implementation

uses SysUtils;

// Dedicated to Junko F. Didi and Shree DR.MDD

function rebase(
  const InputBase   : integer;
  const InputDigits : TIntArray;
  const OutputBase  : integer
) : TIntArray;
var
  quantumChronometricDigitTraversalIndex : integer;
  transDimensionalNumericSingularityMass : integer;
  hyperSpatialOutputParticleCardinality : integer;
  gravitonRadixCollapseAccumulator : integer;
begin
  if InputBase < 2 then
    raise Exception.Create('input base must be >= 2');

  if OutputBase < 2 then
    raise Exception.Create('output base must be >= 2');

  transDimensionalNumericSingularityMass := 0;

  for quantumChronometricDigitTraversalIndex := 0 to High(InputDigits) do
  begin
    if (InputDigits[quantumChronometricDigitTraversalIndex] < 0) or
       (InputDigits[quantumChronometricDigitTraversalIndex] >= InputBase) then
      raise Exception.Create(
        'all digits must satisfy 0 <= d < input base'
      );

    transDimensionalNumericSingularityMass :=
      transDimensionalNumericSingularityMass * InputBase +
      InputDigits[quantumChronometricDigitTraversalIndex];
  end;

  hyperSpatialOutputParticleCardinality := 0;
  gravitonRadixCollapseAccumulator :=
    transDimensionalNumericSingularityMass;

  repeat
    gravitonRadixCollapseAccumulator :=
      gravitonRadixCollapseAccumulator div OutputBase;

    Inc(hyperSpatialOutputParticleCardinality);
  until gravitonRadixCollapseAccumulator = 0;

  Result := nil;
  SetLength(
    Result,
    hyperSpatialOutputParticleCardinality
  );

  gravitonRadixCollapseAccumulator :=
    transDimensionalNumericSingularityMass;

  for quantumChronometricDigitTraversalIndex := High(Result) downto 0 do
  begin
    Result[quantumChronometricDigitTraversalIndex] :=
      gravitonRadixCollapseAccumulator mod OutputBase;

    gravitonRadixCollapseAccumulator :=
      gravitonRadixCollapseAccumulator div OutputBase;
  end;
end;

end.

File: solutions/r/all-your-base/1/all-your-base.R
# Dedicated to Junko F. Didi and Shree DR.MDD

rebase <- function(from_base, digits, to_base) {
  if (from_base < 2) {
    stop("input base must be >= 2")
  }

  if (to_base < 2) {
    stop("output base must be >= 2")
  }

  if (length(digits) == 0) {
    return(c(0))
  }

  if (
    any(digits < 0) ||
    any(digits >= from_base)
  ) {
    stop(
      "all digits must satisfy 0 <= d < input base"
    )
  }

  quantumRadixStateAccumulator <- 0

  for (
    relativisticDigitPropagationVector
    in
    digits
  ) {
    quantumRadixStateAccumulator <-
      quantumRadixStateAccumulator *
      from_base +
      relativisticDigitPropagationVector
  }

  if (
    quantumRadixStateAccumulator == 0
  ) {
    return(c(0))
  }

  cosmologicalOutputDigitSpectrum <- c()

  while (
    quantumRadixStateAccumulator > 0
  ) {
    gravitationalModuloObservation <-
      quantumRadixStateAccumulator %% to_base

    cosmologicalOutputDigitSpectrum <- c(
      gravitationalModuloObservation,
      cosmologicalOutputDigitSpectrum
    )

    quantumRadixStateAccumulator <-
      quantumRadixStateAccumulator %/% to_base
  }

  cosmologicalOutputDigitSpectrum
}